### PR TITLE
Fix VST / HTTP2 idle connection keep alive behaviour

### PR DESCRIPTION
**Pull request by graetzer**
### Scope & Purpose

Make VST and HTTP2 connection have the same keep-alive behaviour as a HTTP 1 connection. Namely only truely cancel an IDLE connection

- [x] :hankey: Bugfix 

#### Backports:

- [x] Backports required for: 3.7

Should be more or less covered by existing tests
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/11663/

### CODE REVIEW

**Comment by graetzer** [comment]
another one with go-tests enabled: http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/11664/

File: arangod/GeneralServer/HttpCommTask.cpp
Comment (maierlars) [inline arangod/GeneralServer/HttpCommTask.cpp:316]: As I understand, this code will always call `asyncReadNextRequest` and thus `_reading` almost always true.

Now, previously I added the `_requestCount` which is now gone again. The reason was the following: imagine the a new request coming in and the timeout expires at the same time. Now both completion handlers are on the queue, first the read completion runs and decides to execute the request, and then restart reading. Now the timeout completion runs, sees that it is still reading and closes the request. Thus we have the same bug as before.

**Review by maierlars — CHANGES_REQUESTED** [review]

File: arangod/GeneralServer/HttpCommTask.cpp
Comment (graetzer) [reply arangod/GeneralServer/HttpCommTask.cpp:316]: I think you are mixing these, this is the http 1.1 version there is no `asyncReadNextRequest` in this class

File: 3rdParty/fuerte/src/GeneralConnection.h
Comment (graetzer) [reply 3rdParty/fuerte/src/GeneralConnection.h:333]: to match what happens on the server, the number is sort of arbitrary

File: arangod/GeneralServer/HttpCommTask.cpp
Comment (graetzer) [reply arangod/GeneralServer/HttpCommTask.cpp:316]: _requestCount was also never increased here

Comment (graetzer) [reply arangod/GeneralServer/HttpCommTask.cpp:316]: VST / HTTP2 now have their own `setIOTimeout()` implementation

Comment (graetzer) [reply arangod/GeneralServer/HttpCommTask.cpp:316]: _reading is not true unless we are onto the next request actually

File: arangod/GeneralServer/H2CommTask.cpp
Comment (fceller) [inline arangod/GeneralServer/H2CommTask.cpp:420]: Should be configurable

Comment (jsteemann) [inline arangod/GeneralServer/H2CommTask.cpp:420]: Why is the timeout at least 300 seconds here? Is there any specific reason for it?

File: arangod/GeneralServer/HttpCommTask.cpp
Comment (jsteemann) [inline arangod/GeneralServer/HttpCommTask.cpp:643]: indentation level looks strange here.

File: arangod/GeneralServer/HttpCommTask.cpp
Comment (jsteemann) [inline arangod/GeneralServer/HttpCommTask.cpp:651]: indentation level looks strange here

File: arangod/GeneralServer/VstCommTask.cpp
Comment (jsteemann) [inline arangod/GeneralServer/VstCommTask.cpp:149]: why is the timeout at least 300 seconds hee? the 300 seconds minimum seems to be in place for HTTP/2 and VST, but not for HTTP/1. I don't see a reason for this.

File: arangod/GeneralServer/H2CommTask.cpp
Comment (graetzer) [reply arangod/GeneralServer/H2CommTask.cpp:420]: because the idle timeout may be super low, so this is a 300 second write timeout

File: arangod/GeneralServer/VstCommTask.cpp
Comment (graetzer) [reply arangod/GeneralServer/VstCommTask.cpp:149]: because i am not gonna change the http 1.1 behaviour

File: arangod/GeneralServer/H2CommTask.cpp
Comment (jsteemann) [reply arangod/GeneralServer/H2CommTask.cpp:420]: If it's a different timeout here (not an idle timeout, but a write timeout), shouldn't it be configurable?

File: arangod/GeneralServer/VstCommTask.cpp
Comment (jsteemann) [reply arangod/GeneralServer/VstCommTask.cpp:149]: Ok, but then this still leaves us with protocol-specific behavior:
* HTTP/1.1 doesn't have a write timeout
* HTTP/2 and VST have a write timeout which is max(idle timeout, 300)

Ideally we wouldn't have differences between the protocols, but try to establish the same behavior for all protocols. If that requires a write-timeout to be configurable, we should add it as a new configuration option.

File: arangod/GeneralServer/HttpCommTask.cpp
Comment (maierlars) [reply arangod/GeneralServer/HttpCommTask.cpp:316]: Got it. I thouht we start reading the next request as soon as we called `processRequest`. However, that is indeed not the case here.

**Review by maierlars — APPROVED** [review]

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/11700/

File: arangod/GeneralServer/VstCommTask.cpp
Comment (graetzer) [reply arangod/GeneralServer/VstCommTask.cpp]: nah some people are sending GB image files as part of foxx apps and stuff like that (and we are not even using sendfile here). I'd rather keep it sane

**Comment by jsteemann** [comment]
LGTM except that the WriteTimeout is still hard-coded to at least 300 seconds. 
I am fine with this being the default value, but I think we should make this value configurable via a startup option.

@fceller @graetzer : ideally we should also make it possible to enforce this write timeout for HTTP/1 connections too, because that would mean all protocols will have the same behavior. 
As that would cause downwards-incompatibility for HTTP/1, this probably needs to be opt-in.
What about having the following options:
* `--http.request-timeout`: config option to set this timeout
* `--http.request-timeout-behavior`: values: "enforce", "compat", with "enforce" enforcing the timeout for all protocols, and "compat" enforcing it just for HTTP/2 and VST (downwards-compatible)

And in a future version, we can deprecate the `--http.*` options and rename them to something else, because they are used for VST as well. Probably something like `--protocol.*`?

**Comment by graetzer** [comment]
i don't know this is just a fix for buggy idle timeouts, i have too many diverging branches, I understand customers are waiting for that

**Comment by jsteemann** [comment]
Has this PR been obsoleted by https://github.com/arangodb/arangodb/pull/12579?
Because https://github.com/arangodb/arangodb/pull/12579 contains the changes made in this PR as well. :thinking:

**Comment by graetzer** [comment]
idgaf if I have to wait for this PR to be merged I cannot work

**Comment by jsteemann** [comment]
@graetzer : I am confused now. 
Is your preference to merge this PR here sooner than https://github.com/arangodb/arangodb/pull/12579, and merge devel into 12579 once we have merged this PR here?
Just not clear to me, so an explanation would help. Thanks!

**Comment by graetzer** [comment]
Whatever

**Pull request merged**
